batcher/aligned-batcher/src/lib.rs

```diff
@@ -192,68 +204,116 @@ impl Batcher {
+                    {
+                        error!("Invalid proof detected. Verification failed.");
+                        send_error_message(
+                            ws_conn_sink.clone(),
+                            ResponseMessage::VerificationError(),
+                        )
+                        .await;
+                        return Ok(()); // Send error message to the client and return
+                    }
+
+                    // Doing nonce verification after proof verification to avoid unnecessary nonce increment
+                    if !self.user_nonce_is_valid(addr, nonce).await {
+                        send_error_message(
+                            ws_conn_sink.clone(),
+                            ResponseMessage::InvalidNonceError,
+                        )
+                        .await;
+                        return Ok(()); // Send error message to the client and return
+                    }
+
+                    self.add_to_batch(
+                        nonced_verification_data,
+                        ws_conn_sink.clone(),
+                        client_msg.signature,
+                    )
                     .await;
-                return Ok(()); // Send error message to the client and return
-            }
+                } else {
+                    error!("Proof is too large");
+                    send_error_message(ws_conn_sink.clone(), ResponseMessage::ProofTooLargeError())
+                        .await;
+                    return Ok(()); // Send error message to the client and return
+                };
 
-            addr
+                info!("Verification data message handled");
+
+                return Ok(());
+            }
         } else {
             error!("Signature verification error");
-            send_error_message(ws_conn_sink.clone(), ResponseMessage::SignatureVerificationError())
-                .await;
-            return Ok(()); // Send error message to the client and return
-        };
+            send_error_message(
+                ws_conn_sink.clone(),
+                ResponseMessage::SignatureVerificationError(),
+            )
+            .await;
+            Ok(()) // Send error message to the client and return
+        }
+    }
 
-        let verification_data = client_msg.verification_data;
-        if verification_data.proof.len() <= self.max_proof_size {
-            // When pre-verification is enabled, batcher will verify proofs for faster feedback with clients
-            if self.pre_verification_is_enabled && !zk_utils::verify(&verification_data) {
-                error!("Invalid proof detected. Verification failed.");
-                send_error_message(ws_conn_sink.clone(), ResponseMessage::VerificationError())
-                    .await;
-                return Ok(()); // Send error message to the client and return
+    async fn user_nonce_is_valid(&self, addr: Address, nonce: U256) -> bool {
```

Comment:
```suggestion
    async fn user_nonce_is_valid(&self, addr: Address, nonce: U256) -> bool {
```
Make it more descriptive and add a comment explaining what to do

```suggestion
    async fn check_user_nonce_is_valid_and_increment_it(&self, addr: Address, nonce: U256) -> bool {
```